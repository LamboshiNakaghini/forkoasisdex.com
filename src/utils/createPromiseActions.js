import { createAction } from 'redux-actions';
import * as storeHelpers from './store';

export function createPromiseActions(actionName,
                                     { pending, fulfilled, rejected } = {}) {
  return {
    'pending': createAction(
      storeHelpers.pending(actionName),
      pending ? pending : (v) => v,
    ),
    'fulfilled': createAction(
      [storeHelpers.fulfilled(actionName)],
      fulfilled ? fulfilled : (v) => v,
    ),
    'rejected': createAction(
      [storeHelpers.rejected(actionName)],
      rejected ? rejected : (v) => v,
    ),
  };
}
