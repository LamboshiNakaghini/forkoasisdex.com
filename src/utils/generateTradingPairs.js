import {
  TOKEN_DAI,
  TOKEN_DIGIX,
  TOKEN_MAKER,
  TOKEN_RHOC,
  TOKEN_WRAPPED_ETH,
  TOKEN_DONUT
} from "../constants";

export const generateTradingPairs = () => {
  return [
    {
      base: TOKEN_WRAPPED_ETH,
      quote: TOKEN_DAI,
      priority: 9,
      isDefault: true
    },
    {
      base: TOKEN_MAKER,
      quote: TOKEN_DAI,
      isDefault: true
    },
    {
      base: TOKEN_MAKER,
      quote: TOKEN_WRAPPED_ETH,
      isDefault: true
    },
    {
      base: TOKEN_DONUT,
      quote: TOKEN_WRAPPED_ETH,
      isDefault: true
    },
    {
      base: TOKEN_DONUT,
      quote: TOKEN_DAI,
      isDefault: false
    },
  ];
};
