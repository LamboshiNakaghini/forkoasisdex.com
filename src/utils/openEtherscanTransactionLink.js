import createEtherscanTransactionLink from './createEtherscanTransactionLink'

function openEtherscanTransactionLink({...kwargs}) {
  window.open(
    createEtherscanTransactionLink(kwargs),
    "_blank",
  );
  window.focus();
}

export default openEtherscanTransactionLink;
