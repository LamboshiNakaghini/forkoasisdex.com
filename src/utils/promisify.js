export default function(func) {
  return (...args) =>
    new Promise((resolve, reject) => {
      const cb = (err, data) => err ? reject(err) : resolve(data);
      func.apply(this, [...args, cb]);
    });
}