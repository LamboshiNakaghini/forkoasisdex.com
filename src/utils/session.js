import sessionReducer from './../store/reducers/session';
import accounts from '../store/selectors/accounts';
import session from '../store/selectors/session';
import { fromJS } from 'immutable';
import web3 from '../bootstrap/web3';

let getState = null;

export const Session = {

  init: (dispatch, getStateFunction) => {
    getState = getStateFunction;
    dispatch(sessionReducer.actions.resetSession());
    const defaultAccountHash = web3.sha3(accounts.defaultAccount(getState()));
    const accountSessionData = sessionStorage.getItem(defaultAccountHash);
    const accountPersistentData = localStorage.getItem(defaultAccountHash);
    if(accountPersistentData) {
      dispatch(
        sessionReducer.actions.loadSavedPersistentData( fromJS(JSON.parse(accountPersistentData))));
    } else {
      localStorage.setItem(defaultAccountHash, JSON.stringify(session.persistentData(getState())));
    }

    if(accountSessionData) {
      dispatch(
        sessionReducer.actions.loadSavedSessionData(fromJS(JSON.parse(accountSessionData)))
      );
    } else {
      sessionStorage.setItem(defaultAccountHash, JSON.stringify(session.sessionData(getState())));
    }
    dispatch(sessionReducer.actions.init());
  },

  getSession: (sessionKey) => getState().getIn(['session', 'session', sessionKey]),
  getPersistent: (sessionKey) => getState().getIn(['session', 'persist', sessionKey]),


  set: (dispatchFunction, key, value) => {
    dispatchFunction(sessionReducer.actions.SetValue(key, value));
    sessionStorage.setItem(
      web3.eth.defaultAccount, JSON.stringify(getState().getIn(['session','session']).toJSON()),
    );
  },
  setPersistent: (dispatchFunction, key, value) => {
    dispatchFunction(sessionReducer.actions.SetValue(key, value));
    localStorage.setItem(
      web3.sha3(web3.eth.defaultAccount), JSON.stringify(getState().getIn(['session', 'persist']).toJSON()),
    );
  },

  dismissMessage: (dispatchFunction, msgType) => {
    dispatchFunction(sessionReducer.actions.dismissMessage(msgType));
    localStorage.setItem(
      web3.sha3(web3.eth.defaultAccount), JSON.stringify(getState().getIn(['session', 'persist']).toJSON())
    );
  },

};