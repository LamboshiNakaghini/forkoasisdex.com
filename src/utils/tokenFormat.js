import BigNumber from 'bignumber.js';

export default function(n, dp = 5) {
  if (isNaN(+n)) return;
  const bn = new BigNumber(n);
  return bn.toFormat(dp);
};