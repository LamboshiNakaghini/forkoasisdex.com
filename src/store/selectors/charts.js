import { createSelector } from 'reselect';
import reselect from '../../utils/reselect';
import moment from 'moment-timezone';
import BigNumber from 'bignumber.js';
import web3 from '../../bootstrap/web3';
import _ from 'lodash';
import {Map, List} from 'immutable';
import {removeOutliersFromArray} from '../../utils/functions';

const trades = state => state.get('trades').get('marketHistory');

const tokenTrades = createSelector(
  trades,
  reselect.getProps,
  (trades, props) => {
    const tokens = props ? [props.tradingPair.baseToken, props.tradingPair.quoteToken] : [];
    return (trades || []).filter(t =>
      t.buyWhichToken == tokens[0] && t.sellWhichToken == tokens[1] ||
      t.buyWhichToken == tokens[1] && t.sellWhichToken == tokens[0]
    )
  },
)


const priceChartTrades = createSelector(
  tokenTrades,
  (trades) => {
    const since = moment(Date.now()).startOf('day').subtract(6, 'days').unix()
    const res = (trades || []).filter(t =>
      t.timestamp >= since
    );
    return res.length == 0 ? [] : res.toJS();
  }
)

const priceChartLabels = createSelector(
  priceChartTrades,
  (priceChartTrades) => priceChartTrades.map((trade) => trade.timestamp)
)

const priceChartValues = createSelector(
  priceChartTrades,
  reselect.getProps,
  (priceChartTrades, props) => priceChartTrades.map((trade) => {
    let baseAmount;
    let quoteAmount;
    if (trade.buyWhichToken === props.tradingPair.quoteToken) {
      quoteAmount = new BigNumber(trade.buyHowMuch);
      baseAmount = new BigNumber(trade.sellHowMuch);
    } else {
      baseAmount = new BigNumber(trade.buyHowMuch);
      quoteAmount = new BigNumber(trade.sellHowMuch);
    }
    return quoteAmount.dividedBy(baseAmount).toFixed(4);
  })
)


const volumeChartTrades = createSelector(
  tokenTrades,
  (trades) => {
    const since = moment(Date.now()).startOf('day').subtract(6, 'days').unix()
    return (trades || []).filter(t =>
      t.timestamp >= since
    );
  }
)

const volumeChartPoints = createSelector(
  () => (
    [6, 5, 4, 3, 2, 1, 0].map(i =>
      moment(Date.now()).startOf('day').subtract(i, 'days')
    )
  )
)

const volumeChartLabels = createSelector(
  volumeChartPoints,
  (volumeChartPoints) => volumeChartPoints.map(d => d.unix())
)

const volumeChartData = createSelector(
  volumeChartTrades,
  volumeChartPoints,
  reselect.getProps,
  (volumeChartTrades, volumeChartPoints, props) => {
      let volumes = {base: {}, quote: {}}
      volumeChartPoints.forEach(day => {
        volumes.base[day.unix()] = new BigNumber(0);
        volumes.quote[day.unix()] = new BigNumber(0);
      })
      volumeChartTrades.forEach(trade => {
        const day = moment.unix(trade.timestamp).startOf('day').unix();
        if (trade.buyWhichToken === props.tradingPair.quoteToken) {
          volumes.quote[day] = volumes.quote[day].add(new BigNumber(trade.buyHowMuch));
          volumes.base[day] = volumes.base[day].add(new BigNumber(trade.sellHowMuch));
        } else {
          volumes.quote[day] = volumes.quote[day].add(new BigNumber(trade.sellHowMuch));
          volumes.base[day] = volumes.base[day].add(new BigNumber(trade.buyHowMuch));
        }
      })
      return volumes;
  }
)

const volumeChartValues = createSelector(
  volumeChartData,
  (volumeChartData) => Object.keys(volumeChartData.quote).map((key) =>
    web3.fromWei(volumeChartData.quote[key]).toFixed(2)
  )
)

const volumeChartTooltips = createSelector(
  volumeChartData,
  (volumeChartData) => ({
    base: Object.values(volumeChartData.base).map(v => web3.fromWei(v).toFormat(2)),
    quote: Object.values(volumeChartData.quote).map(v => web3.fromWei(v).toFormat(2)),
  })
)


const offers = state => state.get('offers').get('offers');

const offersBids = createSelector(
  offers,
  reselect.getProps,
  (offers, props) => (
    !props ? [] : removeOutliersFromArray(
      ((offers.get(Map({baseToken: props.tradingPair.baseToken, quoteToken: props.tradingPair.quoteToken})) || Map()).get('buyOffers') || List()).sortBy(b => b.bid_price_sort).toJS()
    , 'bid_price_sort', 3)
  ),
)

const offersAsks = createSelector(
  offers,
  reselect.getProps,
  (offers, props) => (
    !props ? [] : removeOutliersFromArray(
      ((offers.get(Map({baseToken: props.tradingPair.baseToken, quoteToken: props.tradingPair.quoteToken})) || Map()).get('sellOffers') || List()).sortBy(a => a.ask_price_sort).toJS()
    , 'ask_price_sort', 3)
  ),
)

function upSum(array) {
  return array.reduce(({sum, result}, bn) => (
    {sum: sum.add(bn), result: result.concat(bn.add(sum))}
  ), {sum: new BigNumber(0), result: []}).result;
}

function downSum(array) {
  return array.reduceRight(({sum, result}, bn) => (
    {sum: sum.add(bn), result: [bn.add(sum)].concat(result)}
  ), {sum: new BigNumber(0), result: []}).result;
}

function bigSum(array) {
  return array.reduce((sum, bn) =>
    sum ? sum.add(new BigNumber(bn)) : new BigNumber(bn)
  , null);
}

function groupBy(array, by = _.identity) {
  let result = [];
  array.forEach(e => {
    if ((result[result.length-1] || [])[0] && by((result[result.length-1] || [])[0]) == by(e))
      result[result.length-1].push(e);
    else
      result.push([e]);
  });
  return result;
}

function mapWithPreviousResult(array, mapper = _.identity) {
  return array.reduce(({prevRes, result}, e) => {
    const res = mapper(e, prevRes);
    return {prevRes: res, result: result.concat([res])};
  }, {prevRes: null, result: []}).result;
}

const depthChartData = createSelector(
  offersBids,
  offersAsks,
  (bids, asks) => {
    const askPrices = _.sortedUniq(asks.map(ask => ask.ask_price_sort));
    const askGroups = groupBy(asks, ask => ask.ask_price_sort);
    const askAmounts = {
      quote: upSum(askGroups.map(group => bigSum(group.map(ask => ask.buyHowMuch.toString())))),
      base: upSum(askGroups.map(group => bigSum(group.map(ask => ask.sellHowMuch.toString())))),
    };

    const bidPrices = _.sortedUniq(bids.map(bid => bid.bid_price_sort));
    const bidGroups = groupBy(bids, bid => bid.bid_price_sort);
    const bidAmounts = {
      quote: downSum(bidGroups.map(group => bigSum(group.map(bid => bid.sellHowMuch.toString())))),
      base: downSum(bidGroups.map(group => bigSum(group.map(bid => bid.buyHowMuch.toString())))),
    };

    const vals = _.uniq(bidPrices.concat(askPrices).sort((a, b) => new BigNumber(a.toString()).lt(new BigNumber(b.toString())) ? -1 : 1));

    const askAmountsData = mapWithPreviousResult(vals, (val, prevResult) => {
      const index = askPrices.indexOf(val);
      if (index !== -1) {
        // If there is a specific value for the price in asks, we add it
        return {
          graph: {x: val, y: web3.fromWei(askAmounts.quote[index]).toFixed(3)},
          tooltip: {quote: askAmounts.quote[index], base: askAmounts.base[index]},
        };
      } else if (askPrices.length === 0 ||
        (new BigNumber(val.toString())).lt((new BigNumber(askPrices[0].toString()))) ||
        (new BigNumber(val.toString())).gt((new BigNumber(askPrices[askPrices.length - 1].toString())))) {
        // If the price is lower or higher than the asks range there is not value to print in the graph
        return {
          graph: {x: val, y: null},
          tooltip: {quote: null, base: null},
        };
      } else {
        // If there is not an ask amount for this price, we need to add the previous amount
        return {
          graph: {x: val, y: prevResult.graph.y},
          tooltip: prevResult.tooltip,
        };
      }
    });

    const bidAmountsData = mapWithPreviousResult(vals.slice().reverse(), (val, prevResult) => {
      const index = bidPrices.indexOf(val);
      if (index !== -1) {
        // If there is a specific value for the price in bids, we add it
        return {
          graph: {x: val, y: web3.fromWei(bidAmounts.quote[index]).toFixed(3)},
          tooltip: {quote: bidAmounts.quote[index], base: bidAmounts.base[index]},
        };
      } else if (bidPrices.length === 0 ||
        (new BigNumber(val.toString())).lt((new BigNumber(bidPrices[0].toString()))) ||
        (new BigNumber(val.toString())).gt((new BigNumber(bidPrices[bidPrices.length - 1].toString())))) {
        // If the price is lower or higher than the bids range there is not value to print in the graph
        return {
          graph: {x: val, y: null},
          tooltip: {quote: null, base: null},
        };
      } else {
        // If there is not a bid amount for this price, we need to add the next available amount
        return {
          graph: {x: val, y: prevResult.graph.y},
          tooltip: prevResult.tooltip,
        };
      }
    });

    return {vals, askAmountsData, bidAmountsData: bidAmountsData.reverse()}
  }
)

const depthChartLabels = createSelector(
  depthChartData,
  (depthChartData) => depthChartData.vals
)

const depthChartValues = createSelector(
  depthChartData,
  (depthChartData) => ({
    buy: depthChartData.bidAmountsData.map(d => d.graph),
    sell: depthChartData.askAmountsData.map(d => d.graph),
  })
)

const depthChartTooltips = createSelector(
  depthChartData,
  (depthChartData) => ({
    buy: depthChartData.bidAmountsData.map(v => ({
      base: new BigNumber(web3.fromWei(v.tooltip.base)).toFormat(2),
      quote: new BigNumber(web3.fromWei(v.tooltip.quote)).toFormat(2),
    })),
    sell: depthChartData.askAmountsData.map(v => ({
      base: new BigNumber(web3.fromWei(v.tooltip.base)).toFormat(2),
      quote: new BigNumber(web3.fromWei(v.tooltip.quote)).toFormat(2),
    })),
  })
)


export default {
  priceChartLabels,
  priceChartValues,
  volumeChartLabels,
  volumeChartValues,
  volumeChartTooltips,
  depthChartLabels,
  depthChartValues,
  depthChartTooltips,
}
