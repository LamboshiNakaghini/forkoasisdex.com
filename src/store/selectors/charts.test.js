/* global test expect */
import charts from './charts';
import { Map, List } from 'immutable';
import moment from 'moment-timezone';
import { mockDate, wei } from '../../utils/testHelpers';

const refDate = (f) => mockDate('2018-05-16', f);

const state = refDate(() => Map({
  trades: Map({
    marketHistory: List([
      {
        timestamp: moment('2018-05-16 10:14').unix(),
        buyWhichToken: 'W-ETH',
        sellWhichToken: 'MKR',
        buyHowMuch: wei(200),
        sellHowMuch: wei(100),
      },
      {
        timestamp: moment('2018-05-16 09:14').unix(),
        buyWhichToken: 'W-ETH',
        sellWhichToken: 'SAI',
        buyHowMuch: wei(100),
        sellHowMuch: wei(100),
      },
      {
        timestamp: moment('2018-05-16 09:14').unix(),
        buyWhichToken: 'DAI',
        sellWhichToken: 'MKT',
        buyHowMuch: wei(100),
        sellHowMuch: wei(100),
      },
      {
        timestamp: moment('2018-05-16 08:14').unix(),
        buyWhichToken: 'MKR',
        sellWhichToken: 'W-ETH',
        buyHowMuch: wei(800),
        sellHowMuch: wei(1000),
      },
      {
        timestamp: moment('2018-05-11 08:14').unix(),
        buyWhichToken: 'W-ETH',
        sellWhichToken: 'MKR',
        buyHowMuch: wei(500),
        sellHowMuch: wei(200),
      },
      {
        timestamp: moment('2018-05-10 08:14').unix(),
        buyWhichToken: 'MKR',
        sellWhichToken: 'W-ETH',
        buyHowMuch: wei(200),
        sellHowMuch: wei(500),
      },
      {
        timestamp: moment('2018-05-09 08:14').unix(),
        buyWhichToken: 'MKR',
        sellWhichToken: 'W-ETH',
        buyHowMuch: wei(200),
        sellHowMuch: wei(500),
      },
    ]),
  }),
  offers: Map({
    offers: Map()
      .set(Map({baseToken: 'MKR', quoteToken: 'W-ETH'}), Map({
        buyOffers: List([
          {
            buyHowMuch: wei(900),
            sellHowMuch: wei(1100),
            bid_price: '1.25',
            bid_price_sort: 1.25,
          },
          {
            buyHowMuch: wei(1000),
            sellHowMuch: wei(1200),
            bid_price: '1.27',
            bid_price_sort: 1.27,
          },
        ]),
        sellOffers: List([
          {
            buyHowMuch: wei(250),
            sellHowMuch: wei(200),
            ask_price: '1.28',
            ask_price_sort: 1.28,
          },
          {
            buyHowMuch: wei(300),
            sellHowMuch: wei(250),
            ask_price: '1.26',
            ask_price_sort: 1.26,
          },
        ]),
      })),
  }),
}));

const props = {tradingPair: {baseToken: 'MKR', quoteToken: 'W-ETH'}};


test('priceChartLabels', () => {
  expect(refDate(() => charts.priceChartLabels(state, props))).toMatchSnapshot();
});

test('priceChartValues', () => {
  expect(refDate(() => charts.priceChartValues(state, props))).toMatchSnapshot();
});


test('volumeChartLabels', () => {
  expect(refDate(() => charts.volumeChartLabels(state, props))).toMatchSnapshot();
});

test('volumeChartValues', () => {
  expect(refDate(() => charts.volumeChartValues(state, props))).toMatchSnapshot();
});

test('volumeChartTooltips', () => {
  expect(refDate(() => charts.volumeChartTooltips(state, props))).toMatchSnapshot();
});


test('depthChartLabels', () => {
  expect(refDate(() => charts.depthChartLabels(state, props))).toMatchSnapshot();
});

test('depthChartValues', () => {
  expect(refDate(() => charts.depthChartValues(state, props))).toMatchSnapshot();
});

test('depthChartTooltips', () => {
  expect(refDate(() => charts.depthChartTooltips(state, props))).toMatchSnapshot();
});
