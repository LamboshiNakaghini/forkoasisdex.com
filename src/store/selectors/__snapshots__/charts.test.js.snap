// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`depthChartLabels 1`] = `
Array [
  1.25,
  1.26,
  1.27,
  1.28,
]
`;

exports[`depthChartTooltips 1`] = `
Object {
  "buy": Array [
    Object {
      "base": "1,900.00",
      "quote": "2,300.00",
    },
    Object {
      "base": "1,000.00",
      "quote": "1,200.00",
    },
    Object {
      "base": "1,000.00",
      "quote": "1,200.00",
    },
    Object {
      "base": "0.00",
      "quote": "0.00",
    },
  ],
  "sell": Array [
    Object {
      "base": "0.00",
      "quote": "0.00",
    },
    Object {
      "base": "250.00",
      "quote": "300.00",
    },
    Object {
      "base": "250.00",
      "quote": "300.00",
    },
    Object {
      "base": "450.00",
      "quote": "550.00",
    },
  ],
}
`;

exports[`depthChartValues 1`] = `
Object {
  "buy": Array [
    Object {
      "x": 1.25,
      "y": "2300.000",
    },
    Object {
      "x": 1.26,
      "y": "1200.000",
    },
    Object {
      "x": 1.27,
      "y": "1200.000",
    },
    Object {
      "x": 1.28,
      "y": null,
    },
  ],
  "sell": Array [
    Object {
      "x": 1.25,
      "y": null,
    },
    Object {
      "x": 1.26,
      "y": "300.000",
    },
    Object {
      "x": 1.27,
      "y": "300.000",
    },
    Object {
      "x": 1.28,
      "y": "550.000",
    },
  ],
}
`;

exports[`priceChartLabels 1`] = `
Array [
  1526465640,
  1526458440,
  1526026440,
  1525940040,
]
`;

exports[`priceChartValues 1`] = `
Array [
  "2.0000",
  "1.2500",
  "2.5000",
  "2.5000",
]
`;

exports[`volumeChartLabels 1`] = `
Array [
  1525910400,
  1525996800,
  1526083200,
  1526169600,
  1526256000,
  1526342400,
  1526428800,
]
`;

exports[`volumeChartTooltips 1`] = `
Object {
  "base": Array [
    "200.00",
    "200.00",
    "0.00",
    "0.00",
    "0.00",
    "0.00",
    "900.00",
  ],
  "quote": Array [
    "500.00",
    "500.00",
    "0.00",
    "0.00",
    "0.00",
    "0.00",
    "1,200.00",
  ],
}
`;

exports[`volumeChartValues 1`] = `
Array [
  "500.00",
  "500.00",
  "0.00",
  "0.00",
  "0.00",
  "0.00",
  "1200.00",
]
`;
